x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/AttributeMap.java

```diff
@@ -238,30 +238,52 @@ public boolean isEmpty() {
     @Override
     public boolean containsKey(Object key) {
         if (key instanceof NamedExpression) {
-            return delegate.keySet().contains(new AttributeWrapper(((NamedExpression) key).toAttribute()));
+            return delegate.containsKey(new AttributeWrapper(((NamedExpression) key).toAttribute()));
         }
         return false;
     }
 
     @Override
     public boolean containsValue(Object value) {
-        return delegate.values().contains(value);
+        return delegate.containsValue(value);
     }
 
+    /**
+     * @param key {@link NamedExpression} to look up
+     * @return Looks up the key in the AttributeMap recursively, meaning if the value for this key
+     * is another key in the map, it will follow it. It will return the final value or null if the key
+     * does not exist in the map.
+     */
     @Override
     public E get(Object key) {
-        if (key instanceof NamedExpression) {
-            return delegate.get(new AttributeWrapper(((NamedExpression) key).toAttribute()));
-        }
-        return null;
+        return getOrDefault(key, null);
     }
 
+    /**
+     * @param key {@link NamedExpression} to look up
+     * @return Looks up the key in the AttributeMap recursively, meaning if the value for this key
+     * is another key in the map, it will follow it. It will return the final value or the
+     * specified default value if the key does not exist in the map.
+     */
     @Override
     public E getOrDefault(Object key, E defaultValue) {
-        E e;
-        return (((e = get(key)) != null) || containsKey(key))
-            ? e
-            : defaultValue;
+        E candidate = defaultValue;
+        E value = null;
+        while (((value = lookup(key)) != null || containsKey(key)) && value != key) {
+            key = candidate = value;
+        }
+        return candidate;
+    }
```

Comment:
This is backwards.
getOrDefault should delegate to get not vice-versa. The former is an extension of the latter historically and conceptually.

Moreover by making get and getOrDefault recursive the Map contract has been fundamentally broken:
```
m.put(a, b);
assertTrue(b, m.get(a));
m.put(b,c);
assertTrue(b, m.get(a)); // fails
```

Instead of renaming get to `lookup` and modifying the contract, simply introduce a separate method (e.g. `traverse(key)`, `getResolvedValue(key)`, `getRecursive(key)` or something else) to the original map that does the recursive lookup and use that instead.

Reply:
I think I agree as well on this, I'd prefer to have a dedicated method that does this more complex logic rather than overriding the default `get`.

Reply:
Good points, but maybe the solution here is not just a simple rename of the method (`resolve(key)`), but to not implement the `Map<Attribute, E>` interface anymore. The Map contract was broken since the beginning because instead of `equals()` the `semanticEquals()` is used for key equality checks (which for example violates "containsKey(key) returns true if and only if this map contains a mapping for a key k such that (key==null ? k==null : key.equals(k)).")
After no`get()`s are required anymore, unless you can think of a place where recursive key resolution would actually cause harm (if that is the case, I'd say the AttributeMap building should be fixed instead to limit the scope from where the attributes are picked up from).

Reply:
`AttributeMap` is a `Map` for `AttributeWrapper`s; the `get`, `contains` rely on `equals`. For convenience the underlying class is not exposed instead `Attribute` is used with the wrapping happening automatically.
It's similar to `IdentityHashMap` for example.

`resolve(key)` is a compound method that relies on `get`, the latter might not be used publicly now however I see too many downsides in removing it (and the `Map` contract) instead of keeping it.
For example `get`/`put` do per key read/write but `resolve`, which does multi-key read has not write counter-party; removing/replacing `get` breaks too many assumptions.

Reply:
I am not sure about the downsides or the breaking assumptions, since we barely utilize anything from the Map interface, practically only some read methods: `isEmpty`, `size` and `get/getOrDefault` (which can be renamed to `resolve`). The building of the map (`put`s) happen through the `Builder` methods today, here is how a full minimalization of the public methods would look like (at this point the `AttributeMap` is rather an `AttributeResolver`): https://github.com/palesz/elasticsearch/commit/fe814ec2e0c2a2aa4d082036603c79209c11be6a

Nonetheless, I can keep the current `Map` implementation and create a new `resolve` method for the recursive resolution, switch out all the `get/getOrDefault` calls to `resolve` and leave the `get`s there between the other unused methods.

Reply:
Done